tables_demo_planning/src/tables_demo_planning/tables_demo_api.py

```diff
@@ -1,256 +1,378 @@
-#   * Redistributions in binary form must reproduce the above
-#     copyright notice, this list of conditions and the following
-#     disclaimer in the documentation and/or other materials provided
-#     with the distribution.
-#   * Neither the name of Willow Garage nor the names of its
-#     contributors may be used to endorse or promote products derived
-#     from this software without specific prior written permission.
-#
-#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
-#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
-#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
-#  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
-#  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
-#  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
-#  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
-#  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
-#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
-#  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
-#  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
-#  POSSIBILITY OF SUCH DAMAGE.
-#
-#
-# Authors: Alexander Sung, Oscar Lima, Marc Vinci, Sebastian Stock
-
-"""These Mobipick components implement functionalities in the ROS application domain."""
-
-
-from typing import Dict
 import rospy
-import actionlib
-from std_msgs.msg import String
-from std_srvs.srv import Empty, SetBool
-from geometry_msgs.msg import Pose
-from grasplan.msg import (
-    InsertObjectAction,
-    InsertObjectGoal,
-    PickObjectAction,
-    PickObjectGoal,
-    PlaceObjectAction,
-    PlaceObjectGoal,
-)
-from unified_planning.model import Object
+import rosparam
+
+from typing import Dict, List, Set, Sequence, Union, Optional
+from collections import defaultdict
+from geometry_msgs.msg import Pose, Point
+from unified_planning.plans import ActionInstance
+from tables_demo_planning.components import ArmPose, Item, Location, Robot
+from tables_demo_planning.tables_demo import EnvironmentRepresentation, TablesDemoDomain
+
 from symbolic_fact_generation.on_fact_generator import OnGenerator
-from symbolic_fact_generation.robot_facts_generator import RobotAtGenerator
-from tables_demo_planning.mobipick_components import APIRobot, ArmPose, Item, Location
-from tables_demo_planning.subplan_visualization import SubPlanVisualization
-from tables_demo_planning.tables_demo import TablesDemoDomain, TablesDemoEnv, TablesDemoRobot
+from symbolic_fact_generation.robot_facts_generator import RobotAtGenerator, HasArmPostureGenerator
 from pose_selector.srv import PoseDelete, PoseDeleteRequest
+from mobipick_api import Robot as mobipick_api_robot
+from robot_api import TuplePose, is_instance
+
+
+class TablesDemoAPI:
+    RETRIES_BEFORE_ABORTION = 2
+
+    def __init__(self, item_locations: Dict[Item, Location]) -> None:
```

Comment:
Here I find it unfortunate that `item_locations` needs to be provided. It is only used for providing it as parameter for `EnvironmentRepresentation` where in turn it is also not used in the general case, but only in the textual simulation.
Usually the item locations are not known at the beginning  - especially on the real robot - and if they are they should be queried from the symbolic fact generator. 
To me it looks like this is only for the simulation, in which case it would be better to provide this information only at that specific case.
Furthermore, it is also a bit misleading. Initially I thought this was meant to be prior knowledge that the robot has about the objects.

Reply:
Indeed you are absolutely right. It was only intended for the simulation where the actual item locations are known and needed. For all other files, only the items and (the existence of) locations, not their actual relations are needed. I create a next pull request right away.